src/wallet/wallet.cpp

```diff
@@ -1105,12 +1105,16 @@ void CWallet::transactionAddedToMempool(const CTransactionRef& ptx) {
     }
 }
 
-void CWallet::transactionRemovedFromMempool(const CTransactionRef &ptx) {
+void CWallet::transactionRemovedFromMempool(const CTransactionRef &ptx, bool fIsConflicted) {
     LOCK(cs_wallet);
     auto it = mapWallet.find(ptx->GetHash());
     if (it != mapWallet.end()) {
         it->second.fInMempool = false;
     }
+    if (fIsConflicted) {
+        CWalletTx::Confirmation confirm(CWalletTx::Status::CONFLICTED, /* block_height */ 0, /* block_hash */ {}, /* nIndex */ 0);
```

Comment:
In commit "[rpcwallet] Add `conflicted` as transaction description field" (1af561a8f2bc93ac029b7304fb6da12e6094b860)

I think this should probably be Status::UNCONFIRMED rather than Status::CONFLICTED. That should restore the original behavior before #16624 (a31be09bfd77eed497a8e251d31358e16e2f2eb1 +   7e89994133725125dddbfa8d45484e3b9ed51c6e) and avoid introducing an entirely new transaction state that wallet code has never had to deal with, where a transaction is conflicting with something not in a specific block. This should also allow reverting `GetDepthInMainChain` below and leaving it unchanged.

At a high level I think it makes sense to mark these transactions unconfirmed instead of conflicted because we know there are cases the wallet can't detect unconfirmed transactions becoming conflicted, and cases where it can't detect a conflicting transactions becoming unconfirmed, and unconfirmed is the more conservative choice than conflicted for balances, etc

I haven't looked at your `wallet_txn_doublespend.py` test yet, so I'm not sure how changing this would interact that. Another way of testing this code which might be better from the point of view of #18600 is to add a regression test to `feature_notifications.py` to ensure that this commit restores the `-walletnotify` notifications that were lost in #16624

Reply:
Thanks for your comment, I didn't reverse to the previous behavior starting from the user viewpoint. Unconfirmed funds allow you to spend from then or consider likely-yours as part of your balance, but here we know, as of mempool state, they are conflicted and likely-but-not-certain not going to confirm. We don't want the user to take decisions based on funds he has not. You can spend from unconfirmed, right ? 

Knowing there are cases the wallet can't detect unconfirmed transactions becoming conflicted doesn't mean that when we have the information available we shouldn't act on this.

Thanks for `feature_notifications.py`, we will modify to test notifications, didn't know its existence.

Reply:
> avoid introducing an entirely new transaction state that wallet code has never had to deal with, where a transaction is conflicting with something not in a specific block.

Ah sorry reading this again, in fact in `transactionRemovedFromMempool` you may have case of transaction block conflicted but not covered by `BlockConnected` due to the conflicting transaction not involving us. 

Like unconfirmed txA spent by unconfirmed txB, txB paying a wallet address. txA is conflicted by txA' in a block, we won't see it in BlockConnected right now.

So this PR restores UI notifications and _extend_ scope of block conflicted transactions in the lightweight way I think. Detecting them at block connection is not possible without tracking whole chain of unconfirmed transactions AFAICT.

> Is this going to work correctly when m_confirm.hashBlock is null, or does it need to be updated too?

Child conflicted transactions is going to inherit `m_confirm.hashBlock` null from parent.

>  If GetDepthInMainChain and LoadToWallet have to be updated to deal with the new state, and these transactions are serialized to wallet files, what happens if the wallets are loaded by older wallet software?

It's not a new state, as `Confirmation::Status::CONFLICTED" is already there. I think what matters is among `CONFLICTED` transactions is there any code path relying on `hashBlock` or `block_height`, beyond display code, there is `MarkConflicted` to keep track of the most conflicting code and `GetDepthInMainChain` ofc. 

> Will 0.20 and 0.19 GetDepthInMainChain and LoadToWallet implementations handle the new state correctly without bugs and crashes?

As said previously, it's not a new state, and you already have of course `UNCONFIRMED` with null hash and zero-height on which current serialization/deserialization logic relies on : https://github.com/bitcoin/bitcoin/blob/b6a5dc90bfd4640cf9f914e59bf8e21cd265b51e/src/wallet/wallet.h#L434

Trying to go forward, I think best would be to pass block height and hash in `transactionsRemovedFromMempool` even if it  seems counter-intuitive, for block conflicting transaction we do have a block height/hash. 

What do you think ?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/18600#discussion_r412587833

> It's not a new state

The new state which this writes is:

| `uint256 hashBlock` | `int nIndex` | interpretation                             |
|---------------------|--------------|--------------------------------------------|
| 0                   | -1           | conflicted since #16624, unconfirmed prior |

The existing states written are:

| `uint256 hashBlock` | `int nIndex` | interpretation                             |
|---------------------|--------------|--------------------------------------------|
| 0                   | 0            | unconfirmed                                |
| >1                  | >=0          | confirmed                                  |
| >1                  | -1           | conflicted since #7105, unconfirmed prior  |
| 1                   | -1           | abandoned since #7312, unconfirmed prior   |

Never written states and still unused states are:

| `uint256 hashBlock` | `int nIndex` | interpretation                             |
|---------------------|--------------|--------------------------------------------|
| 0                   | not 0 or -1  | unconfirmed                                |
| 1                   | not -1       | abandoned since #7312, unconfirmed prior   |
| >1                  | <-1          | unconfirmed                                |

Maybe it's good to add a new state, or maybe it's bad, but it doesn't seem like enough thought has gone into it the decision so far. Looking at `CWalletTx::GetDepthInMainChain`, `CWallet::ComputeTimeSmart`, `CWallet::MarkConflicted` all of these methods seem broken by the new state. If I am reading correctly, GetDepthInMainChain will return large bogus negative height values, MarkConflicted 'more conflicted' check will start overwriting valid conflicted blocks with null conflict blocks, ComputeTimeSmart will logging spurious found in block not in index errors. Am I misreading the code in seeing these behaviors? Or do you see the same behaviors, but think side effects are minimal? Do you think existing 0.17, 0.18, 0.19 wallet code handles transactions in the new state safely?

> Trying to go forward, I think best would be to pass block height and hash in `transactionsRemovedFromMempool`

This does seem like a safer way forward, and probably a good approach. Only caveat is the usual caveat with changing the wallet conflicted transaction heuristic: that if we start marking transactions conflicted instead of unconfirmed, the wallet can fail to notice them becoming unconfirmed again if they become unconflicted later. This should be less likely when the conflict is coming from a block transaction instead of a mempool transaction, but not certain. I think if we are going to change the heuristic, we can't just think about it abstractly. We should be looking at specific cases and trying to figure out if users are helped or hurt by the change in each case.

Reply:
You're right this approach by introducing a new state is broken. I think I've been confused by trying to solve #18325 and at same time try to harden conflict detection on non-connected wallet transaction.

I think the safe approach is to pass block height and hash in `transactionsRemovedFromMempool` but in another PR to reason on its own.

Note: IIRC we don't have undo conflict logic, that was the whole intent of https://github.com/bitcoin/bitcoin/pull/17543.